{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import urllib.request\n",
    "import gzip\n",
    "import qgrid\n",
    "import gffutils\n",
    "import sqlite3\n",
    "\n",
    "from itertools import repeat\n",
    "from scipy.spatial.distance import pdist\n",
    "\n",
    "import os.path as osp\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# download data here if it doesn't exist already\n",
    "net_file_path = './dm3.droVir3.net.gz'\n",
    "net_file_url = 'http://hgdownload.soe.ucsc.edu/goldenPath/dm3/vsDroVir3/dm3.droVir3.net.gz'\n",
    "\n",
    "if not osp.exists(net_file_path):\n",
    "    print('downloading synteny file')\n",
    "    urllib.request.urlretrieve(net_file_url, net_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Net File Format Reference: https://genome.ucsc.edu/goldenPath/help/net.html\n",
    "\n",
    "data = []\n",
    "\n",
    "def StringIsInt(s):\n",
    "    try: \n",
    "        int(s)\n",
    "        return True\n",
    "    except ValueError:\n",
    "        return False\n",
    "\n",
    "\n",
    "with gzip.open(net_file_path, 'r') as netfile:\n",
    "    for i, line in enumerate(netfile):\n",
    "        fields = line.decode('utf-8').strip().split()\n",
    "        \n",
    "        if fields[0] == 'net':\n",
    "            current_chromosome = fields[1]\n",
    "            continue\n",
    "        \n",
    "        # lines prefixed with 'gap' are not matching\n",
    "        lineIsNotSynteny = fields[0] != 'fill'\n",
    "        \n",
    "        if lineIsNotSynteny:\n",
    "            continue\n",
    "    \n",
    "        # These are attributes that are mandatory and fixed    \n",
    "        _, tstart, tsize, qname, plusminus, qstart, qsize = fields[:7]\n",
    "        \n",
    "        # There are extra attributes that are not always specified/are optional\n",
    "        optional_attributes = {}\n",
    "\n",
    "\n",
    "        for optional_attribute_name in fields[7::2]:\n",
    "            index_of_attribute = fields.index(optional_attribute_name)\n",
    "            attribute_value = fields[index_of_attribute + 1]\n",
    "            optional_attributes[optional_attribute_name] = int(attribute_value) \\\n",
    "                                                            if StringIsInt(attribute_value)\\\n",
    "                                                            else attribute_value\n",
    "        \n",
    "        attributes_table  = {\n",
    "                 'melano_chromosome': current_chromosome, \n",
    "                 'melano_start': int(tstart),\n",
    "                 'melano_size': int(tsize),\n",
    "                 'virilis_chromosome': qname,\n",
    "                 'relative_orientation': plusminus,\n",
    "                 'virilis_start': int(qstart),\n",
    "                 'virilis_size': int(qsize)}\n",
    "        \n",
    "        \n",
    "        attributes_table.update(optional_attributes)\n",
    "        \n",
    "        data.append(attributes_table)\n",
    "\n",
    "        \n",
    "my_data_frame = pd.DataFrame(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "qgrid.show_grid(my_data_frame)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "syntenic_query = 'melano_size > 10000 and virilis_size > 10000 and type == \"syn\" '\\\n",
    "                 'and melano_chromosome == \"chr2L\"'\n",
    "\n",
    "big_syntenic_blocks = my_data_frame.query(syntenic_query)\\\n",
    "    .sort_values('score', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "if not osp.exists('syntenic_blocks.csv'):\n",
    "    print('saving syntenic blocks to csv')\n",
    "    big_syntanic_blocks.to_csv('syntenic_blocks.csv')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "qgrid.show_grid(big_syntenic_blocks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# long range contact stuff\n",
    "long_range_contact_url = 'http://www.cell.com/cms/attachment/2007964571/2030739776/mmc3.xls'\n",
    "long_range_contact_path = './long_range_contacts.xls'\n",
    "\n",
    "if not osp.exists(long_range_contact_path):\n",
    "    print('downloading long range contacts')\n",
    "    urllib.request.urlretrieve(long_range_contact_url, long_range_contact_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "long_range_contacts = pd.read_excel(long_range_contact_path, header=3)\n",
    "qgrid.show_grid(long_range_contacts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# on 2l, the contact points\n",
    "chromosome_2L_points = long_range_contacts.query('Chr == \"2L\"')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.4/site-packages/ipykernel/__main__.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "# looped areas (should count from outside of sticking area)\n",
    "start_stop_points = ['Xstart', 'Xend', 'Ystart', 'Yend']\n",
    "forward_loops = chromosome_2L_points['Yend'] - chromosome_2L_points['Xstart']\n",
    "backward_loops = chromosome_2L_points['Xend'] - chromosome_2L_points['Ystart']\n",
    "\n",
    "is_forward_loop = forward_loops > 0\n",
    "is_backward_loop = ~is_forward_loop\n",
    "\n",
    "loop_lengths = forward_loops * is_forward_loop + backward_loops * is_backward_loop\n",
    "chromosome_2L_points['loop_length'] = loop_lengths\n",
    "qgrid.show_grid(chromosome_2L_points)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "filtered_points = chromosome_2L_points.query('loop_length < 3e6') \n",
    "qgrid.show_grid(filtered_points)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "num_loops:  30\n",
      "loop coverage of chromosome 2L: 0.206956521739\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.4/site-packages/matplotlib/__init__.py:892: UserWarning: axes.color_cycle is deprecated and replaced with axes.prop_cycle; please use the latter.\n",
      "  warnings.warn(self.msg_depr % (key, alt_key))\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<matplotlib.text.Text at 0x116a66278>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAERCAYAAAB8eMxzAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAEhhJREFUeJzt3X2QXXV9x/F3sgs0sAGTdIlAIZEMfDulI1YceSjlSVAB\nO4rWYWqRFkSqZVDGKWpkULFSHR9QwYdRnutjBQtWVOgDFsFqHCxTpeIXaFxwqIYlu0BigpBk+8c5\ni0uyu7kJe85J8nu/Znbm3nPv3d93zjn72d/9nXN+Z9bY2BiSpHLM7roASVK7DH5JKozBL0mFMfgl\nqTAGvyQVxuCXpML0N/WLI6IfuApYDOwMXAz8FLgG2ADcnZnnNNW+JGlyTfb4TwMeycyjgJcDnwQu\nAd6VmUcDsyPilQ22L0maRJPB/1XgwvpxH7AOeGFm3l4v+zZwfIPtS5Im0dhQT2auAYiIucB1wAXA\nRya8ZRWwR1PtS5Im1+jB3YjYF7gVuDYzv0I1tj9uLvBok+1LkjbV5MHdhcAtwDmZ+Z168V0RcVRm\nfhc4keqfwrTWrVs/1t/fB8C9997LXZddwaIFg02VPakHVg7zR+eexYEHHthqu5L0LMya6oXGgh9Y\nCjwHuDAi3g2MAW8FLouInYB7gOs390tGR9c8/XhkZDWLFgyyZOFezVQ8jZGR1QwPr2q9XUnaGoOD\nc6d8rckx/vOA8yZ56Zim2pQkbZ4XcElSYQx+SSqMwS9JhTH4JakwBr8kFcbgl6TCGPySVBiDX5IK\nY/BLUmEMfkkqjMEvSYUx+CWpMAa/JBXG4Jekwhj8klQYg1+SCmPwS1JhDH5JKozBL0mFMfglqTAG\nvyQVxuCXpMIY/JJUGINfkgpj8EtSYQx+SSqMwS9JhTH4JakwBr8kFcbgl6TCGPySVBiDX5IKY/BL\nUmEMfkkqjMEvSYUx+CWpMAa/JBXG4Jekwhj8klQYg1+SCmPwS1JhDH5JKozBL0mFMfglqTAGvyQV\nxuCXpMIY/JJUGINfkgrT33QDEXEo8MHMPDYiXgDcBNxbv/yZzLyu6RokSb/VaPBHxPnA64HV9aJD\ngI9m5seabFeSNLWmh3ruB06Z8PwQ4OSIuC0iroiI3RpuX5K0kUaDPzNvANZNWLQMOD8zjwaWA+9t\nsn1J0qYaH+PfyI2Z+Vj9+Abg0s19YN68Xenv7wNgdHSAkQaLm878+QMMDs7tqHVJmjltB//NEXFu\nZt4JvAT40eY+MDq65unHIyOrp3lns0ZGVjM8vKqz9iVpS0zXUW07+N8EfCoingR+BZzdcvuSVLzG\ngz8zHwCOqB//N3Bk021KkqbmBVySVBiDX5IKY/BLUmEMfkkqjMEvSYUx+CWpMAa/JBXG4Jekwhj8\nklQYg1+SCmPwS1JhDH5JKozBL0mFMfglqTAGvyQVxuCXpMIY/JJUGINfkgpj8EtSYQx+SSqMwS9J\nhTH4JakwBr8kFcbgl6TCGPySVBiDX5IKY/BLUmEMfkkqjMEvSYXp77oANWP9+vUMDS1vvd3Fi/en\nr6+v9XYl9c7g30ENDS3na9e9kcHBOa21OTy8lte89nKWLDmgtTYlbTmDfwc2ODiHvZ67W9dlSNrG\nOMYvSYUx+CWpMAa/JBXG4Jekwhj8klQYg1+SCmPwS1JhDH5JKozBL0mF6Sn4I+KySZZdO/PlSJKa\nNu2UDRFxBbA/8KKIOGjCSzsBezRZmCSpGZubq+f9wGLgE8BFE5avA+5pqCZJUoOmDf7MHAKGgIMj\nYneqXv6s+uUBYKTJ4iRJM6+n2TkjYimwFFg5YfEY1TCQJGk70uu0zGcBSzJzuMliJEnN6/V0zgdx\nWEeSdgi99vjvA+6IiO8AT4wvzMz3be6DEXEo8MHMPDYilgDXABuAuzPznC0vWZL0bPTa438IuBn4\nDdXB3fGfaUXE+cDlwC71okuAd2Xm0cDsiHjlFlcsSXpWeurxZ+ZFm3/XpO4HTgE+Xz8/JDNvrx9/\nGzgB+PpW/m5J0lbo9ayeDVRn8Uz0f5m573Sfy8wbImLRhEUTvyWswovAJKl1vfb4nx4SioidgFcB\nh29FexsmPJ4LPLq5D8ybtyv9/X0AjI4OdHaEef78AQYH53bU+pYbHR3opN3tbT1JJer14O7TMvMp\n4LqIuGAr2vuviDgqM78LnAjcurkPjI6uefrxyMjqrWhyZoyMrGZ4eFVn7W+prtbV9raepB3VdB2w\nXod6Tp/wdBZwEPDkVtTyt8Dl9beGe4Drt+J3SJKehV57/MdOeDwGPAKc2ssHM/MB4Ij68X3AMVtQ\nnyRphvU6xn9G3UuP+jN3Z+a6RiuTJDWi1/n4D6G6iOta4GrgwfrCLEnSdqbXoZ5LgVMzcxlARBwG\nXAa8uKnCJEnN6PXK3YHx0AfIzB8Av9NMSZKkJvUa/CMTp1eIiFfxzCmaJUnbiV6Hes4GboqIK6lO\n5xyjPlNHkrR96bXHfyKwBlhEdWrnI3hapiRtl3oN/rOBP87MX2fmj4EXAuc2V5YkqSm9Bv9OPPNK\n3SfZdNI2SdJ2oNcx/huBWyPiq/XzV+N0ypK0Xeqpx5+Z76A6lz+obrB+aWZe2GRhkqRm9Dw7Z2Ze\nj5OqSdJ2r9cxfknSDsLgl6TCGPySVBiDX5IKY/BLUmEMfkkqjMEvSYUx+CWpMAa/JBXG4Jekwhj8\nklQYg1+SCmPwS1JhDH5JKozBL0mFMfglqTAGvyQVpuc7cGly69evZ2hoeSdtL168P319fZ20LWn7\nZfA/S0NDy7nri3/DPgt2bbXdh1augb/4NEuWHNBqu5K2fwb/DNhnwa4sXjjQdRmS1BPH+CWpMAa/\nJBXG4Jekwhj8klQYg1+SCmPwS1JhDH5JKozBL0mFMfglqTAGvyQVxuCXpMIY/JJUGINfkgpj8EtS\nYQx+SSqMwS9JhenkRiwR8SPgsfrpzzPzDV3UIUklaj34I2IXgMw8ru22JUnd9PgPBnaLiFuAPuCC\nzFzWQR2SVKQuxvjXAB/OzJcBbwa+GBEea5CklnTR478XuB8gM++LiJXAXsBDk7153rxd6e/vA2B0\ndICRtqrcyPz5AwwOzt1k+ejoAA93UA9MXRNUdXVhupokbRu6CP4zgOcD50TE3sBc4JdTvXl0dM3T\nj0dGVjde3FRGRlYzPLxq0uVdmaqm8de6MF1NktozXQesi+C/ErgqIr4LjAFnZuaGDuqQpCK1HvyZ\nuQ44ve12JUkVD6pKUmEMfkkqjMEvSYUx+CWpMAa/JBXG4Jekwhj8klQYg1+SCmPwS1JhDH5JKozB\nL0mFMfglqTAGvyQVxuCXpMIY/JJUGINfkgpj8EtSYQx+SSqMwS9JhTH4JakwBr8kFcbgl6TCGPyS\nVBiDX5IKY/BLUmEMfkkqjMEvSYUx+CWpMP1dFyB1af369QwNLe+k7cWL96evr6+TtlU2g19FGxpa\nzlu+eRVzFi5otd21K1Zy6clnsmTJAa22K4HBLzFn4QIG9l7YdRlSaxzjl6TCGPySVBiDX5IKY/BL\nUmEMfkkqjMEvSYUx+CWpMAa/JBXG4JekwnjlrlrjvDi92RbX07ZYE3RX17ZYE/S+nxv8as3Q0HLe\n/a2zGVg4p9V2V69Yy/tO+tx2My/O0NByzrvpJnbds91pJNY8vIKPv+IVk66noaHlfOBbP2X3hfu2\nWtPjK37B0pOYctsNDS3nP792H3sNLmqtpl8OPwCvmb6moauXsd/8fVqrCeDBkYfgjKnrmsjgV6sG\nFs5hj71367qMbd6uey5kYO92g2Nzdl+4L/P23r/rMjax1+Ai9nvukq7LeIb95u/Dkj0Xd13GlBzj\nl6TCGPySVBiDX5IKY/BLUmFaP7gbEbOATwMHA08AZ2VmN+c+SVKBuujxvwrYJTOPAJYCl3RQgyQV\nq4vgPxK4GSAzlwEv6qAGSSpWF8G/O/DYhOfrIsJjDZLUki4u4HocmDvh+ezM3NDrhx9YOTzzFfXQ\n5nTXBT60ck1rtUxsc8/NvGd4eG0rtWxJe6tXtFtTL22uXbGypUp6b3PNwytaqqT3Nh9f8YuWKtm4\nzT+Y9j2/HH6gnWImtPc8pr869sGRh1qq5pltLub3enrvrLGxsYbLeaaIeDXwisw8MyIOAy7MzJNb\nLUKSCtZFj/8G4ISI+F79/IwOapCkYrXe45ckdcuDqpJUGINfkgpj8EtSYQx+SSrMDn8jlog4FPhg\nZh4bES8ALgXWAb8BTs/M1i8M2KimLwMLgVnAYuD7mfm6FmuZBVwBBLAeOBt4b1c1TbK9bgLurV/+\nTGZeFxFvA04D1gKfzMwvd1DXJvtRRLyRav09BVycmd9soa5+4Cqq7bQzcDHwA+By4DlU2/D0zGzt\nZPf6gszLqfapDcCbqNbTNfXzuzPznBbqmGzdPAhcxqbb7s3AmXV9H8jMG1uua8ptVv+NfhO4MTM/\nNxM17NA9/og4n2pl7lIv+jhwTmYeR3Va6Tu7rikz/7yu5xRgFDiv5ZJeCuyWmUcCf0cVWJ3UNMn2\nOgT4aGYeV/9cFxEHAa8HDgWOAy6IiM1dyzbTdW28H70jIhYC5wKHAy8HPhAROzVZV+004JHMPAo4\nEfgk8CHgC5l5DPAe4A9bqGOiPwXG6n3qQuDvqebkeldmHg3MjohXtlDHxHXzcqp1M9m22w04HzgM\neFn9nrbq6mWbvZ/qH8KM2aGDH7ifKrzGnZqZP6kf91P1GNu2cU3jLgIuy8yHW67nCWCPulexB/Bk\nhzVtvG4OAU6OiNsi4vKIGKC6jPM/MvOpzPwNcDfVH2ybdW28Hz0BvBi4IzPXZebjwH3A8xuuC+Cr\nVOEK1d/zOuAIYN+I+FfgdcCtLdTxtMz8OtU3H4BFVJ2HF2bm7fWybwPHt1DKxHXTR/VNbLJtN1b/\nzAUGqL75tlXXtNssIl5T13PzTBawQwd/Zt5AtVLHn68AiIgjgHOAj3VdU13PIFXv9Zq26wHuAOYA\nPwM+SzWE0UlNk6ybZcD5dS/x51Q9oZ8AR0XEbhGxgOoPptGb+Pa4H208B9Vqqn+kjcrMNZn564iY\nC1wHXAA8D1iZmScAv6CDb7aZuSEirqban75ENXwxbhUdrZvxTszEbZeZa4CvAD8F7qxrbrUuJtlm\n9bfb11Ht97Om/IVbYYcO/slExKlU9wM4KTPbn6Rlcn8GfCkzu7ia7u3A9zIzqO6R8A8RsXPHNY27\nMTPvqh/fALwgM38GfIqqB3Qp1djoI20XNsl+9DhV+I+bCzzaUi37UvUQr83Mr1Ctj2/UL3+D6ptT\n6zLzDOBAqmNIcya81NW6+cd62TO2XUQcTjVEtwjYDzglIhqdNbiHbfYiqiHNvev3/RXwtoh46Uy0\nX1TwR8RpVP/lj2nzYNcUJv4HP57q628XBvhtT/VRqq+/fR3XNO7mCX+ALwF+FBG/C+yemX8CvJlq\n6OcHbRY1xX70Q+DIiNg5IvYAfp9qGKrpWhYCtwBvz8xr68V3ACfVj48C/qfpOjaq6fURsbR++gTV\nUMWdEXF0vexE4PZJPzyzdWyybqbYdgPAmnr48Emqv4MZHVPfXF1sus3uzsx3ZubhmXks1TfvSzLz\nX2aihh3+rJ5x9ZkGnwAeAG6IiDHgtsy8qKOSJvakDwS6ugvZh4GrI+J2qv1haWaujYguaxr3JuBT\nEfEk8Cvg7MxcHZUfUgXK2zNzVVsFTbcfRcSlVH/As6gOZD45za+aKUupQurCiHg31X71l8CV9Zkq\nj1ENF7TpeuCaiLiNap96C9VQ4hX1Ae976vc0beN10wccxOTb7oSIWEY1pHdHZv5bi3W1vs2cq0eS\nClPUUI8kyeCXpOIY/JJUGINfkgpj8EtSYQx+SSqMwS9NIiKOjojvdF2H1ASDX5qaF7loh1TMlbvS\n1oiIA4DPAfOpJl57a2beWU8FfSXV3C5PUU0AdktEvAdYQnU19gLgs5n5kW6qlyZnj1+a2izgC8DH\nM/Ng4G3A9fW0A5cB/14vfy1wVT2jKVTzBx1DNdHWX9c3bpG2GQa/NLUBYEk9vzyZuQxYSTUB23FU\nPX4y8+dUE8UdWn/u85n5RGY+Bvxz/V5pm2HwS1ObzabzoM+mGiKdajk8854Cszd6LnXO4Jem9jjw\nvxFxCkBEHEZ1L+KfUM2Rfla9fH+qG8J8v/7cayNip4iYR3UbwhmZSleaKQa/NLUxqvujviUifkx1\n45dTMnMd8FbguHr5PwFvGL8zF9Uc9N+rfy6ubx4jbTOcllmaQfVZPWsz80Nd1yJNxR6/JBXGHr8k\nFcYevyQVxuCXpMIY/JJUGINfkgpj8EtSYQx+SSrM/wMahDZxsi7yGQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x1169d2128>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "loops = pd.DataFrame()\n",
    "\n",
    "loops['loop_start'] = filtered_points[start_stop_points].min(axis=1)\n",
    "loops['loop_stop'] = filtered_points[start_stop_points].max(axis=1)\n",
    "\n",
    "# def overlap(start1, end1, start2, end2):\n",
    "#     \"\"\"Does the range (start1, end1) overlap with (start2, end2)?\"\"\"\n",
    "#     return end1 >= start2 and end2 >= start1\n",
    "\n",
    "def overlap(a,b,c,d):\n",
    "    r = 0 if a==c and b==d else min(b,d)-max(a,c)\n",
    "    if r >= 0: \n",
    "        return r\n",
    "    else:\n",
    "        return 0\n",
    "\n",
    "def overlapper(m_start, m_end):\n",
    "    def overlap_inner(row):\n",
    "        return overlap(row['loop_start'], row['loop_stop'], m_start, m_end)\n",
    "    \n",
    "    return overlap_inner\n",
    "\n",
    "def array_overlap(segment_a, segment_b):\n",
    "    return overlap(segment_a[0], segment_a[1], segment_b[0], segment_b[1])\n",
    "    \n",
    "\n",
    "blocks = dict(zip(big_syntenic_blocks.index, repeat(None)))\n",
    "for row in big_syntenic_blocks.itertuples():\n",
    "    # calculate coverage with loops\n",
    "    index = row[0]\n",
    "    melano_size = row[4]\n",
    "    melano_start = row[5]\n",
    "    melano_end = melano_start + melano_size\n",
    "    \n",
    "    overlap_function = overlapper(melano_start, melano_end)\n",
    "    \n",
    "    coverage = loops.apply(overlap_function, axis=1) / melano_size\n",
    "    \n",
    "    covering_loop = coverage.argmax()\n",
    "    \n",
    "    blocks[index] = covering_loop\n",
    "    #     print('block {} coverage {} covering loop {}'.format(index, coverage.max(), covering_loop))\n",
    "    \n",
    "block_series = pd.Series(blocks)\n",
    "    \n",
    "# how many loops are there\n",
    "num_loops = len(loops)\n",
    "\n",
    "print('num_loops: ', num_loops)\n",
    "\n",
    "# how much of this chromosome is covered by loops\n",
    "# you can calculate this by generating the condensed overlap matrix\n",
    "all_overlaps = pdist(loops[['loop_start', 'loop_stop']].values, array_overlap).sum()\n",
    "all_loops = (loops['loop_stop'] - loops['loop_start']).sum()\n",
    "\n",
    "unique_loop_coverage = all_loops - all_overlaps\n",
    "unique_loop_ratio = unique_loop_coverage / all_loops\n",
    "\n",
    "chr_2l_length = 23e6\n",
    "\n",
    "print('loop coverage of chromosome 2L:', unique_loop_coverage / chr_2l_length)\n",
    "\n",
    "# what is the distribution of loops covered by each\n",
    "# print(block_series)\n",
    "sns.countplot(block_series)\n",
    "plt.xlabel('loop')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# insulator data\n",
    "# download from http://gbrowse.modencode.org/fgb2/gbrowse/fly/?label=Insulators_CLASSES_KW for class 1, 2\n",
    "\n",
    "insulator_file = './Insulators_CLASSES_KW_2L.gff3.txt'\n",
    "\n",
    "insulator_db_fn = './insulators.db'\n",
    "\n",
    "if not osp.exists(insulator_db_fn):\n",
    "    gffutils.create_db(insulator_file, insulator_db_fn)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# read insulator db into a pandas frame\n",
    "with sqlite3.connect(insulator_db_fn) as connection:\n",
    "    insulator_df = pd.read_sql_query(\"SELECT * from features\", connection)\n",
    "    insulator_df.drop(['score', 'strand', 'frame', 'extra'], axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# we want to see that insulator "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
